### PR TITLE
Resolves #21: New testframework

### LINKED ISSUE #21: Use different framework for assertions in unittests
Methods like `assertEquals(a, b)` are confusing and might easily be used the wrong way (which parameter is the expected one? When adding a message, does it come first or last?).
There are other frameworks that provide more readable assertions like `assertThat(myActualValue).isEqualTo(myExpectedValue)`. My personal favourite is [Truth](https://truth.dev/) but there might be good alternatives.
Another option would be to write unittests more "kotlin style", options woth checking might be:
- Spek: https://www.spekframework.org
- KotlinTest (looks promising!): https://github.com/kotlintest/kotlintest
- Other stuff from [this list](https://kotlin.link/?q=test)?

### PR DESCRIPTION
**Pull request by Baret**
#21
A new framwork to write more elegant unittests.

I will first go with https://github.com/kotlintest/kotlintest/blob/master/doc/reference.md and see how it feels.

### CODE REVIEW

**Comment by Baret** [comment]
For reference, those tests should at least be executed after this issue is done:
```
[INFO] Running de.gleex.pltcmd.model.world.WorldTileTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.068 s - in de.gleex.pltcmd.model.world.WorldTileTest
[INFO] Running de.gleex.pltcmd.model.world.CoordinateTest
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.029 s - in de.gleex.pltcmd.model.world.CoordinateTest
[INFO] Running de.gleex.pltcmd.model.world.SectorTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.081 s - in de.gleex.pltcmd.model.world.SectorTest
[INFO] Running de.gleex.pltcmd.model.world.WorldMapTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.256 s - in de.gleex.pltcmd.model.world.WorldMapTest
[INFO] Running de.gleex.pltcmd.model.elements.ArmyElementHierarchyTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.095 s - in de.gleex.pltcmd.model.elements.ArmyElementHierarchyTest
[INFO] Running de.gleex.pltcmd.model.radio.RadioSignalTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.043 s - in de.gleex.pltcmd.model.radio.RadioSignalTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 15, Failures: 0, Errors: 0, Skipped: 0
```

**Comment by Baret** [comment]
I put a quick summary of my findings [in issue 21](https://github.com/Baret/pltcmd/issues/21#issuecomment-580749970).
Happy review :)

**Comment by Baret** [comment]
Tests running currently:
```
2020-01-31T13:04:06.8133246Z [INFO] Running de.gleex.pltcmd.model.elements.ArmyElementHierarchyTest
2020-01-31T13:04:07.2087118Z [INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.736 s - in de.gleex.pltcmd.model.elements.ArmyElementHierarchyTest
2020-01-31T13:04:07.2095080Z [INFO] Running de.gleex.pltcmd.model.world.CoordinateTest
2020-01-31T13:04:07.8033819Z [INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.585 s - in de.gleex.pltcmd.model.world.CoordinateTest
2020-01-31T13:04:07.8035187Z [INFO] Running de.gleex.pltcmd.model.world.MainCoordinateTest
2020-01-31T13:04:08.0600566Z [INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.249 s - in de.gleex.pltcmd.model.world.MainCoordinateTest
2020-01-31T13:04:08.0618488Z [INFO] Running de.gleex.pltcmd.model.radio.RadioSignalTest
2020-01-31T13:04:08.9145589Z [WARNING] Tests run: 15, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.851 s - in de.gleex.pltcmd.model.radio.RadioSignalTest
2020-01-31T13:04:08.9176738Z [INFO] Running de.gleex.pltcmd.model.world.SectorTest
2020-01-31T13:04:09.1696268Z [INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.238 s - in de.gleex.pltcmd.model.world.SectorTest
2020-01-31T13:04:09.1756961Z [INFO] Running de.gleex.pltcmd.model.world.WorldMapTest
2020-01-31T13:04:11.6283316Z [INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.44 s - in de.gleex.pltcmd.model.world.WorldMapTest
2020-01-31T13:04:11.6289126Z [INFO] Running de.gleex.pltcmd.model.world.WorldTileTest
2020-01-31T13:04:11.8272857Z [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.199 s - in de.gleex.pltcmd.model.world.WorldTileTest
2020-01-31T13:04:12.1581832Z [INFO] 
2020-01-31T13:04:12.1582636Z [INFO] Results:
2020-01-31T13:04:12.1582956Z [INFO] 
2020-01-31T13:04:12.1583391Z [WARNING] Tests run: 46, Failures: 0, Errors: 0, Skipped: 1
```

File: src/test/kotlin/io/kotlintest/provided/ProjectConfig.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            override fun beforeSpecClass(spec: Spec, tests: List<TopLevelTest>) {
+                log.info("Starting tests ${spec.description().fullName()}")
+            }
+
+            override fun afterSpecClass(spec: Spec, results: Map<TestCase, TestResult>) {
+                logMemoryUsage()
+            }
+
+            override fun beforeTest(testCase: TestCase) {
+                Runtime.getRuntime().gc()
```
Comment (Loomie) [inline src/test/kotlin/io/kotlintest/provided/ProjectConfig.kt]: The GC slows down the tests! The JVM will run it when necessary. But the used heap is not that meaningful then.

File: src/main/kotlin/de/gleex/pltcmd/model/terrain/TerrainHeight.kt [hunk 2]
```diff
@@ -1,6 +1,9 @@
 package de.gleex.pltcmd.model.terrain
 
-enum class TerrainHeight(val value: Int) {
+/**
+ * Represents the height of a tile.
+ */
+enum class TerrainHeight(val value: Int): Comparable<TerrainHeight> {
```
[the remark this answers is not in the document]
Comment (Baret) [reply src/main/kotlin/de/gleex/pltcmd/model/terrain/TerrainHeight.kt]: The default compares by the ordinal value (order in which it appears in the enum). To not rely on the cleverness of the developer filling he enum... ;)

File: src/main/kotlin/de/gleex/pltcmd/model/world/MainCoordinate.kt [hunk 3]
```diff
@@ -10,4 +10,6 @@ data class MainCoordinate(val eastingFromLeft: Int, val northingFromBottom: Int)
      * Converts this main coordinate o a [Coordinate]
      */
     fun toCoordinate() = Coordinate(eastingFromLeft * 100, northingFromBottom * 100)
+
+    override fun toString() = "($eastingFromLeft|$northingFromBottom)"
```
[the remark this answers is not in the document]
Comment (Baret) [reply src/main/kotlin/de/gleex/pltcmd/model/world/MainCoordinate.kt]: What is a "Coordinates text"?

File: src/test/kotlin/de/gleex/pltcmd/model/elements/ArmyElementHierarchyTest.kt [hunk 4]
```diff
@@ -1,67 +1,144 @@
...
+import io.kotlintest.matchers.types.beNull
+import io.kotlintest.matchers.types.shouldBeSameInstanceAs
+import io.kotlintest.should
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.FreeSpec
 
-class ArmyElementHierarchyTest {
+class ArmyElementHierarchyTest: FreeSpec({
+    val division = ArmyElementHierarchy.Division.createElement()
+    "A division called $division should" - {
```
[the remark this answers is not in the document]
Comment (Baret) [reply src/test/kotlin/de/gleex/pltcmd/model/elements/ArmyElementHierarchyTest.kt]: Valid point!

File: src/test/kotlin/io/kotlintest/provided/ProjectConfig.kt (hunk 1 above)
Comment (Baret) [reply src/test/kotlin/io/kotlintest/provided/ProjectConfig.kt]: Yeah it was more a test to see if it helps reducing the needed heap space as the tests were using too much memory.

File: src/main/kotlin/de/gleex/pltcmd/model/terrain/Terrain.kt [hunk 5]
```diff
@@ -1,3 +1,23 @@
 package de.gleex.pltcmd.model.terrain
 
-data class Terrain(val type: TerrainType, val height: TerrainHeight)
\ No newline at end of file
+class Terrain private constructor(val type: TerrainType, val height: TerrainHeight) {
```
[the remark this answers is not in the document]
Comment (Baret) [reply src/main/kotlin/de/gleex/pltcmd/model/terrain/Terrain.kt]: An IDE warning told me, that it has a generated copy() function, which would make the constructor more accessible or something. I then thought, that I do not want any copies of Terrain objects. That's why I removed the "data".

File: src/main/kotlin/de/gleex/pltcmd/model/world/MainCoordinate.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (Loomie) [reply src/main/kotlin/de/gleex/pltcmd/model/world/MainCoordinate.kt]: the result of Coordinate.toString()

[the remark this answers is not in the document]
Comment (Baret) [reply src/main/kotlin/de/gleex/pltcmd/model/world/MainCoordinate.kt]: Oh, I see. I dind't realize the comment was at the MainCoordinte :D

File: src/test/kotlin/de/gleex/pltcmd/model/world/CoordinateTest.kt [hunk 6]
```diff
@@ -99,5 +100,22 @@ class CoordinateTest: WordSpec({
         "be $expectedString for $testCoordinate" {
             testCoordinate.toString() shouldBe expectedString
         }
+
+        "be padded for single digits" {
+            toCoordinateString(1, 1) shouldBe "(001|001)"
+            toCoordinateString(2, 9) shouldBe "(002|009)"
+            toCoordinateString(-1, -9) shouldBe "(-01|-09)"
```
Comment (Baret) [inline src/test/kotlin/de/gleex/pltcmd/model/world/CoordinateTest.kt]: This is wrong. Coordinates should always have 3 digits in each part! I will fix that.

Comment (Baret) [reply src/test/kotlin/de/gleex/pltcmd/model/world/CoordinateTest.kt]: I tried to fix this, but the toString is rather complicated now. But I couln't find a better implementation...

File: src/main/kotlin/de/gleex/pltcmd/model/world/Coordinate.kt [hunk 7]
```diff
@@ -31,6 +31,22 @@ data class Coordinate(val eastingFromLeft: Int, val northingFromBottom: Int) : C
...
 
-    override fun toString() = "(${"%03d".format(eastingFromLeft)}|${"%03d".format(northingFromBottom)})"
+    override fun toString(): String {
+        val eastingString = if(eastingFromLeft > 0) {
+            FORMAT_POSITIVE.format(eastingFromLeft)
+        } else {
+            FORMAT_NEGATIVE.format(eastingFromLeft)
+        }
+        val northingString = if(northingFromBottom > 0) {
+            FORMAT_POSITIVE.format(northingFromBottom)
```
Comment (Loomie) [inline src/main/kotlin/de/gleex/pltcmd/model/world/Coordinate.kt]: Don't Repeat Yourself ([DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself))

File: src/main/kotlin/de/gleex/pltcmd/model/world/Coordinate.kt [hunk 8]
```diff
@@ -31,6 +31,22 @@ data class Coordinate(val eastingFromLeft: Int, val northingFromBottom: Int) : C
...
     }
 
-    override fun toString() = "(${"%03d".format(eastingFromLeft)}|${"%03d".format(northingFromBottom)})"
+    override fun toString(): String {
+        val eastingString = if(eastingFromLeft > 0) {
+            FORMAT_POSITIVE.format(eastingFromLeft)
+        } else {
+            FORMAT_NEGATIVE.format(eastingFromLeft)
+        }
+        val northingString = if(northingFromBottom > 0) {
```
Comment (Loomie) [inline src/main/kotlin/de/gleex/pltcmd/model/world/Coordinate.kt]: equal to zero should also be positive!

**Pull request merged**